source/common/stats/thread_local_store.cc

```diff
@@ -204,29 +255,130 @@ Histogram& ThreadLocalStoreImpl::ScopeImpl::histogram(const std::string& name) {
   // See comments in counter(). There is no super clean way (via templates or otherwise) to
   // share this code so I'm leaving it largely duplicated for now.
   std::string final_name = prefix_ + name;
-  HistogramSharedPtr* tls_ref = nullptr;
+  std::unique_lock<std::mutex> lock(parent_.lock_);
+  ParentHistogramImplSharedPtr& central_ref = central_cache_.histograms_[final_name];
```

Comment:
Why not have a thread-local cache for parents that we check before going to the central cache?  There are two ways we could do this:
1.  Store a reference to the parent in the tls histogram, try to grab the tls histogram from its thread-local cache, and grab the parent from it if it exists.
2.  Create a thread-local cache just for the parent histograms that is totally distinct from the thread-local cache for tls histograms.

Reply:
Added TLS cache via #2. Please check once.